File: app/lib/models/chat_message.dart
enum ChatMessageType {
  text,
  image,
  video;

  static ChatMessageType fromString(String? val) {
    switch (val) {
      case 'image':
        return ChatMessageType.image;
      case 'video':
        return ChatMessageType.video;
      case 'text':
      default:
        return ChatMessageType.text;
    }
  }

  String get value {
    switch (this) {
      case ChatMessageType.image:
        return 'image';
      case ChatMessageType.video:
        return 'video';
      case ChatMessageType.text:
        return 'text';
    }
  }
}

/// Model representing a chat message in a private hopping squad.
class ChatMessage {
  const ChatMessage({
    required this.id,
    required this.squadId,
    required this.senderId,
    required this.senderName,
    required this.timestamp,
    this.senderPhotoUrl,
    this.type = ChatMessageType.text,
    this.text,
    this.mediaUrl,
    this.thumbnailUrl,
  });

  final String id;
  final String squadId;
  final String senderId;
  final String senderName;
  final String? senderPhotoUrl;
  final ChatMessageType type;
  final String? text;
  final String? mediaUrl;
  final String? thumbnailUrl;
  final DateTime timestamp;

  bool isUser(String currentUserId) => senderId == currentUserId;
  bool get isVideo => type == ChatMessageType.video;
  bool get isImage => type == ChatMessageType.image;

  String get senderInitials {
    final parts = senderName.trim().split(' ');
    if (parts.length >= 2) {
      return '${parts[0][0]}${parts[1][0]}'.toUpperCase();
    }
    return senderName.isNotEmpty ? senderName[0].toUpperCase() : 'H';
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'squadId': squadId,
        'senderId': senderId,
        'senderName': senderName,
        'senderPhotoUrl': senderPhotoUrl,
        'type': type.value,
        'text': text,
        'mediaUrl': mediaUrl,
        'media_url': mediaUrl,
        'thumbnailUrl': thumbnailUrl,
        'timestamp': timestamp.millisecondsSinceEpoch,
      };

  factory ChatMessage.fromJson(Map<String, dynamic> json) {
    DateTime ts;
    final rawTs = json['timestamp'];
    if (rawTs is int) {
      ts = DateTime.fromMillisecondsSinceEpoch(rawTs);
    } else if (rawTs is String) {
      ts = DateTime.tryParse(rawTs) ?? DateTime.now();
    } else {
      ts = DateTime.now();
    }

    return ChatMessage(
      id: json['id'] as String? ?? 'msg_${DateTime.now().millisecondsSinceEpoch}',
      squadId: json['squadId'] as String? ?? '',
      senderId: json['senderId'] as String? ?? 'unknown',
      senderName: json['senderName'] as String? ?? 'Hopper',
      senderPhotoUrl: json['senderPhotoUrl'] as String? ?? json['sender_photo_url'] as String?,
      type: ChatMessageType.fromString(json['type'] as String?),
      text: json['text'] as String?,
      mediaUrl: json['mediaUrl'] as String? ?? json['media_url'] as String?,
      thumbnailUrl: json['thumbnailUrl'] as String? ?? json['thumbnail_url'] as String?,
      timestamp: ts,
    );
  }
}

File: ios/Uma/Models/ChatMessage.swift
// ChatMessage.swift — text-only for v1

import Foundation

enum ChatMessageType: String, Codable, Sendable {
    case text, image, video

    static func from(_ s: String?) -> ChatMessageType {
        switch s {
        case "image": return .image
        case "video": return .video
        default:      return .text
        }
    }
}

struct ChatMessage: Identifiable, Hashable, Codable, Sendable {
    let id: String
    let squadId: String
    let senderId: String
    let senderName: String
    let senderPhotoUrl: String?
    let type: ChatMessageType
    let text: String?
    let mediaUrl: String?
    let thumbnailUrl: String?
    let timestamp: Date

    func isUser(_ uid: String) -> Bool { senderId == uid }
    var isVideo: Bool { type == .video }
    var isImage: Bool { type == .image }

    var senderInitials: String {
        let parts = senderName.split(separator: " ").map(String.init)
        if parts.count >= 2 { return "\(parts[0].prefix(1))\(parts[1].prefix(1))".uppercased() }
        return senderName.isEmpty ? "H" : String(senderName.prefix(1)).uppercased()
    }

    enum CodingKeys: String, CodingKey {
        case id, squadId, senderId, senderName, senderPhotoUrl
        case type, text, mediaUrl, thumbnailUrl, timestamp
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = (try? c.decode(String.self, forKey: .id))
            ?? "msg_\(Int(Date().timeIntervalSince1970 * 1000))"
        squadId = (try? c.decode(String.self, forKey: .squadId)) ?? ""
        senderId = (try? c.decode(String.self, forKey: .senderId)) ?? "unknown"
        senderName = (try? c.decode(String.self, forKey: .senderName)) ?? "Hopper"
        senderPhotoUrl = try? c.decode(String.self, forKey: .senderPhotoUrl)
        type = ChatMessageType.from(try? c.decode(String.self, forKey: .type))
        text = try? c.decode(String.self, forKey: .text)
        mediaUrl = try? c.decode(String.self, forKey: .mediaUrl)
        thumbnailUrl = try? c.decode(String.self, forKey: .thumbnailUrl)
        if let ts = try? c.decode(Double.self, forKey: .timestamp) {
            timestamp = Date(timeIntervalSince1970: ts / 1000.0)
        } else if let s = try? c.decode(String.self, forKey: .timestamp),
                  let d = ISO8601DateFormatter().date(from: s) {
            timestamp = d
        } else {
            timestamp = Date()
        }
    }

    init(id: String, squadId: String, senderId: String, senderName: String,
         senderPhotoUrl: String? = nil, type: ChatMessageType = .text,
         text: String? = nil, mediaUrl: String? = nil, thumbnailUrl: String? = nil,
         timestamp: Date = Date()) {
        self.id = id; self.squadId = squadId
        self.senderId = senderId; self.senderName = senderName
        self.senderPhotoUrl = senderPhotoUrl; self.type = type
        self.text = text; self.mediaUrl = mediaUrl; self.thumbnailUrl = thumbnailUrl
        self.timestamp = timestamp
    }
}
